library/core/src/num/dec2flt/common.rs

```diff
@@ -0,0 +1,234 @@
+//! Common utilities, for internal use only.
+
+use crate::marker::PhantomData;
+use crate::ptr;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct AsciiStr<'a> {
+    ptr: *const u8,
+    end: *const u8,
+    _marker: PhantomData<&'a [u8]>,
```

Comment:
Why does this need to use raw pointers? Couldn't you wrap a &[u8]?

Reply:
I could, but to be honest, I didn't change it because it was part of the original implementation and wasn't broken. As long as `step_by` is implemented properly (it shouldn't need redundant bounds-checks), it would also make `check_len` and the rest of the methods a bit simpler logic-wise.

I'll look into implementing those changes today, running benchmarks, and ensuring every safety guarantee still holds. Thanks for the feedback.

Reply:
A quick check shows that using slice itself might be more [efficient](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlJvoCqAZ0wAFAB4gA5AAYppAFZdSrZrVDEArkIKkR7ZATx1KmWugDCqVpoC2tZWfQAZPLUwA5OwCNMxEABsAJykAA6oQoTGtFY29srhkUZ0ru5etr7%2BwXqYBsm0DATMxASxdg6cOXnRhcUEqZ4%2BfoEhOnVl8ZVtJQ3pmS0AlHqomsTIHFIApABMAMxgYJMArABCbKyoAO4QmrTamOgA%2BrtCzFSYA8sAIpMyAIK3d/sA1DroICC2xQDWzSCiCDURFsV2YRUmsxWj0eoU03leBC0hmeADU/JE6JwIRYwFJmBDsM9JgB2KH3Z4U56hREgZ4AKjQtB0z00AA5SI9Kc8nLSGXRmWyOeTKYcvsRfv5ngCgagQWD8bMLDMArjmETVmzrgShQ9iTd7o88LZQqxsaqCaj0dEsYrzbNCSSyXcubD4VRaM93NshLTlar1StNUsrgNngBaC0MXJUImkzlcylR1gxx3xhMJ6mSoQAOmYQkOmYgQzT6cpAHoy88GHcAGLYAAqAE1aU5nrZtARnr5nm5ngQEJhXsxbIPisRmABPZ55qmI7Ml0vc8y0k5nQeO165/OFga59DoCA59i0IuhklXHWLrmin7NKWA%2Biy0FFD5YKjMTSsAhFy%2Bl88l/8DWFCkK0rO50AANzUcY%2BwHZ4ILwTBNi7KdaFIadIOgtxgB7Ts3DDE1mBgohnggWhs2zXcS1dFkmTXZ53QRTBQkObwJwgZV207fQqHQhwWUiAAvC5wwtTjNE7JMUzjYCEwrKtawbZshxHachCYli2PQ05zjU54NnUT0e3U9ghHU/s1FgwdvE0KhzmIfSzGAfsF0pHjs0zIlZiuWidPXUlXmjDy52YfcyLPPUISdRceIAyKgOdct5PAqDaGIuCEKQlDPXQ0LUuQbDcJ7WgCNUYjUFIzgKKo2SaNXXTGJ0ZiOOmAIuMC5NQwje0iVa9qpNjaLS3k6s6ybWk/L0gycJnHjjM9VBO0wY0CAnVyKXq/yVg6qhsyazT2M4CL9USilAIeWTQOeABxTA8JjftByEVg8BgxkijcdTaAq3JlrMAgc2ouEGI9PB82W6l2OVHiuotbxUGsQb1p24KHIhG5vJRpw4pO86YWBnQkU7NFiAxWhpjNPExJk07XlYZBfVa/1lkDVlrh1PH7iNE1KYVQkSbJinbSpnrU1q4HGK9Q96cZlU8QDIMQ1EnqBrF2muVVgLnuQWMTq5TnaaulLoMHR74MQ5C2JyjD8sKwhitKojTYqsjqvncX4U2kGNNYqG%2Bokna%2BJXISRO6h1/ck6MkdkrkRsU8aVNHdT9t97T6JnaajLBxyzNgyyzZsuy/Ec9QXJjtygu1ryfK9jd3O17NgDu440oHZBfgPcjKN1qLkdi2SDdj5LMLS02Mot7K0JtrDDPt/DCPKyq3aBz26Iaj19patqA5h5Xw%2B3yPk2j9WkoUsblMmjO6Bm5Oo%2Bz76lpWtby42tetpRlO2IgI6e4H%2BKLsNvJW690rJ01eoOd6zBPoLW5OwEc9BAYe29mDQ4ENVpb13mHLsCNWDHxipXem2YUFoPYpcP%2BuN/74zdB6L4bgixI3/lIIYrBpBLHkA4WQ8hUDSAAEodleCMMY645icHkAQaQcgBhDG%2BCAJYMgVDSG4OwiRXDpDyB9PI8RnChhwFgEgNAxo8DsDIBQCABjQhGOaMQTgSxZjyKoEYgg6JKDeBUaQbwbhigTmkKI0gBj4EEAAPK0FYN4zhpAsBfHUOwNx%2BBiCYEMHgCCmAfThMwOIBJEkJi%2BLcE4lh4SXreHHMQCcVgsA%2BLEcQI0FShj8EYCwGJPA%2BB0AIMIMQkhwlKEqKodQIAtA6BUHgbwPpIBDFQNSaIqSwwWGeOMggYZ2DJNNN5WY6jfqJJMBAJwHQKikCcL0Jo/hKiJCiHQHZCQIinNoAcjIzQujrPyLUEo5z7nVDoE8%2BobhGi3KOXoIozzrDlGUN0T5aRDlcCGEIQR4wIUKKkGw0gHC5CkG4VIfpBAdZVVsdmGQpFcCEBIL1WYlRnhWEMcYolP8%2BE6DESoqRpAZFyLhUoxFbjUXqJAJoulzDpDTGUeE9lpAtGSKGMk0m0QQDcCAA), anyway. So, this is probably a worthwhile fix outside of just code duplication.

```asm
;; Version1 uses the primitive slice type.
example::Version1::new:
        mov     rax, rdi
        lea     rdx, [rsi + rdi]
        ret

example::Version1::step_by:
        mov     rax, rdi
        add     qword ptr [rdi], rsi
        ret

example::Version1::step:
        mov     rax, rdi
        add     qword ptr [rdi], 1
        ret

example::Version1::is_empty:
        mov     rax, qword ptr [rdi]
        cmp     rax, qword ptr [rdi + 8]
        sete    al
        ret

;; Version2 uses the the raw pointers.
example::Version2::new:
        mov     rdx, rsi
        mov     rax, rdi
        ret

example::Version2::step_by:
        mov     rax, rdi
        add     qword ptr [rdi], rsi
        sub     qword ptr [rdi + 8], rsi
        ret

example::Version2::step:
        mov     rax, rdi
        add     qword ptr [rdi], 1
        add     qword ptr [rdi + 8], -1
        ret

example::Version2::is_empty:
        cmp     qword ptr [rdi + 8], 0
        sete    al
        ret

example::main:
        ret
```

Reply:
This has been implemented as of https://github.com/Alexhuszagh/rust/commit/ba7f2bc5733914658de6295924491bd044db12bc.

Most of the relevant functionality has been added to `ByteSlice`, and `AsciiStr` now uses a slice primitive [internally](https://github.com/Alexhuszagh/rust/blob/ba7f2bc5733914658de6295924491bd044db12bc/library/core/src/num/dec2flt/common.rs#L139-L143):

```rust
/// Bytes wrapper with specialized methods for ASCII characters.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AsciiStr<'a> {
    slc: &'a [u8],
}
```

Reply:
I believe this is resolved, let me know if there are any further issues.